feat: add `lapack/base/dlas2`

Pranavchiku opened the pull request:
Towards #2464.

## Description

> What is the purpose of this pull request?

This pull request adds JS implementation for `lapack/base/dlas2`.

## Related Issues

> Does this pull request have any related issues?

NA

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Pranavchiku committed:
feat: init base implementation

Pranavchiku committed:
docs: add readme

Pranavchiku committed:
docs: add repl

Pranavchiku committed:
feat: add other files

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlas2/README.md:
```diff
@@ -0,0 +1,221 @@
+var Float64Array = require( '@stdlib/array/float64' );
+var dlas2 = require( '@stdlib/lapack/base/dlas2' );
+
+var out = new Float64Array( 2 );
+dlas2( 1.0, 2.0, 3.0, out );
+console.log( out );
```
@Pranavchiku You need to go beyond just copy-pasting a prior example. This applies to all your PRs.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlas2/benchmark/benchmark.js:
```diff
@@ -0,0 +1,60 @@
+	var i;
+
+	b.tic();
+	A = ( randu() * 100.0 ) - 50.0;
+	B = ( randu() * 100.0 ) - 50.0;
+	C = ( randu() * 100.0 ) - 50.0;
+	out = new Float64Array( 2 );
```
Why are you doing this within the `b.tic()`? This benchmark needs to be refactored. It is also not clear why were are reusing `out` values in subsequent iterations.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlas2/benchmark/benchmark.ndarray.js:
```diff
@@ -0,0 +1,61 @@
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		out = dlas2( A, B, C, out, 1, 0 );
+		A = out[ i%out.length ];
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlas2/lib/base.js:
```diff
@@ -0,0 +1,106 @@
+		if ( fhmx === 0.0 ) {
+			ssmax = ga;
+		} else {
+			ssmax = max( fhmx, ga ) * sqrt( 1.0 + pow( min( fhmx, ga ) / max( fhmx, ga ), 2 ) );
```
For squaring numbers, use `abs2`. It is a much lighter weight dependency and will be faster in general.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlas2/lib/base.js:
```diff
@@ -0,0 +1,106 @@
+		} else {
+			as = 1.0 + ( fhmn / fhmx );
+			at = ( fhmx - fhmn ) / fhmx;
+			c = 1.0 / ( sqrt( 1.0 + pow( as * au, 2 ) ) + sqrt( 1.0 + pow( at * au, 2 ) ) );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlas2/test/test.dlas2.js:
```diff
@@ -0,0 +1,114 @@
+
+	out = new Float64Array( 2 );
+	out = dlas2( 2.0, 3.0, 4.0, out );
+	expected = new Float64Array( [ 1.5513263285176897, 5.1568776039816795 ] );
```
@Pranavchiku How did you compute these expected values? R, Mathematica, something else? It would be good to include a comment here (as we do in other parts of the codebase) specifying how expected values were computed.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlas2/test/test.dlas2.js:
```diff
@@ -0,0 +1,114 @@
+	t.end();
+});
+
+tape( 'the function computes the eigenvalues of a 2x2 symmetric matrix', function test( t ) {
```
I believe you are computing singular values, not eigenvalues values. Applies here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlas2/test/test.ndarray.js:
```diff
@@ -0,0 +1,92 @@
+	t.end();
+});
+
+tape( 'the function supports complex access pattern to store computed values', function test( t ) {
```
@Pranavchiku You need to include the tests from `test.dlas2.js` here. Don't take shortcuts. It is not a given that the underlying implementation is the same for the main and ndarray exports. Always test to the interface, NOT the implementation.

kgryte replied:
Having such a test is fine ("complex access patterns"), but you also should have tests for specific behavior. E.g., "the function supports an `out` stride", "the function supports supports an `out` offset", etc, where you hold everything constant except for the behavior you want to test. Tests should build on one another.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlas2/test/test.ndarray.js:
```diff
@@ -0,0 +1,92 @@
+	t.end();
+});
+
+tape( 'the function supports accessing elements in reverse order to store computed values', function test( t ) {
```
```suggestion
tape( 'the function supports a negative `out` stride', function test( t ) {
```

Follow the conventions found elsewhere for test descriptions.

prajjwalbajpai committed:
Merge branch 'develop' into dlas2

prajjwalbajpai committed:
refactor: benchmarks, tests and update docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---
